internal/ruletree/ruletree_test.go

```diff
@@ -0,0 +1,243 @@
+		rawLists = append(rawLists, data)
+	}
+	b.SetBytes(totalBytes)
+
+	for b.Loop() {
+		tree := ruletree.NewRuleTree[*spyData]()
+		for _, data := range rawLists {
+			scanner := bufio.NewScanner(bytes.NewReader(data))
+
+			for scanner.Scan() {
+				line := scanner.Text()
+				if line == "" {
+					continue
+				}
+
+				if err := tree.Add(line, &spyData{}); err != nil {
+					b.Fatalf("add rule %q: %v", line, err)
+				}
+			}
+
+			if err := scanner.Err(); err != nil {
+				b.Fatalf("scan: %v", err)
+			}
+		}
+	}
+
+	b.ReportAllocs()
+}
+
+func BenchmarkMatch(b *testing.B) {
+	tree, err := loadTree()
+	if err != nil {
+		b.Fatalf("load tree: %v", err)
+	}
+
+	reqs, avgBytes, err := loadReqs()
+	if err != nil {
+		b.Fatalf("load reqs: %v", err)
+	}
+	b.SetBytes(avgBytes)
+
+	var i int
+	for b.Loop() {
+		r := reqs[i]
+		i++
+		if i == len(reqs) {
+			i = 0
+		}
+		tree.FindMatchingRulesReq(r)
+	}
+
+	b.ReportAllocs()
+}
+
+func BenchmarkMatchParallel(b *testing.B) {
+	tree, err := loadTree()
+	if err != nil {
+		b.Fatalf("load tree: %v", err)
+	}
+
+	reqs, avgBytes, err := loadReqs()
+	if err != nil {
+		b.Fatalf("load reqs: %v", err)
+	}
+	b.SetBytes(avgBytes)
+
+	b.ResetTimer()
+
+	var seedCounter int64
+	b.RunParallel(func(pb *testing.PB) {
+		id := atomic.AddInt64(&seedCounter, 1)
+		r := rand.New(rand.NewSource(baseSeed + id)) // #nosec G404 -- Not used for cryptographic purposes.
+
+		i := r.Intn(len(reqs))
+		for pb.Next() {
+			r := reqs[i]
+			i++
+			if i == len(reqs) {
+				i = 0
+			}
+			tree.FindMatchingRulesReq(r)
+		}
+	})
+
+	b.ReportAllocs()
+}
+
+func loadTree() (*ruletree.RuleTree[*spyData], error) {
+	tree := ruletree.NewRuleTree[*spyData]()
+
+	for _, filename := range filterLists {
+		data, err := os.ReadFile(filename)
+		if err != nil {
+			return nil, fmt.Errorf("read %s: %v", filename, err)
+		}
+
+		scanner := bufio.NewScanner(bytes.NewReader(data))
+		for scanner.Scan() {
+			line := scanner.Text()
+			if line == "" {
+				continue
+			}
+
+			err := tree.Add(line, &spyData{})
+			if err != nil {
+				return nil, fmt.Errorf("add rule %q: %v", line, err)
+			}
+		}
+
+		if err := scanner.Err(); err != nil {
+			return nil, fmt.Errorf("scan %s: %v", filename, err)
+		}
+	}
+	return tree, nil
+}
+
+// loadReqs generates a list of HTTP requests from URLs in testdata/urls.txt
+// and synthetic URLs. It returns the requests, average URL length in bytes,
+// and any error encountered.
+func loadReqs() ([]*http.Request, int64, error) {
+	urls, err := loadURLs()
+	if err != nil {
+		return nil, 0, err
+	}
+
+	reqs := make([]*http.Request, len(urls))
+	var totalURLBytes int64
+	for i, u := range urls {
+		reqs[i] = &http.Request{
+			Method: http.MethodGet,
+			URL:    u,
+			Host:   u.Hostname(),
+		}
+		totalURLBytes += int64(len(u.String()))
+	}
+	avg := totalURLBytes / int64(len(urls))
+
+	// Shuffle the elements to avoid ordering bias.
+	rnd.Shuffle(len(reqs), func(i, j int) {
+		reqs[i], reqs[j] = reqs[j], reqs[i]
+	})
```

Comment:
This is nice, with module this can yield very stable and good benchmark results. I like this.